src/9FC02700.c

```diff
@@ -1,27 +1,423 @@
 #include "include_asm.h"
+#include "bbtypes.h"
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", func_9FC02700);
+extern const char aRoot[];
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", func_9FC02734);
+// TODO: rename these
+extern const u8 pubkey[];
+extern const u32 exponent;
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", get_cert_body);
+extern const char aXS[];
+extern const char aCP[];
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", verify_cert_signature);
+extern BbVirage01 D_9FC0F308;
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", verify_cert_chain);
+extern const char aRootCpca[];
+extern const char aRootXsca[];
+extern const char aRoot_0[];
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", get_expected_revocation_list_version);
+s32 cert_get_size(BbCertBase* certBase) {
+	switch(certBase->certType) {
+        case 0:
+            return sizeof(BbEccCert);
+        case 1:
+            return sizeof(BbRsaCert);
+        default:
+            return -1;
+    }
+}
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", check_crl_root);
+s32 cert_get_signature_size(BbCertBase* certBase) {
+    switch(certBase->sigType) {
+        case 0:
+        case 1:
+            return sizeof(BbRsaSig4096);
+        case 2:
+            return sizeof(BbEccSig);
+        default:
+            return -1;
+    }
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", check_crlbundle_version);
+}
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", verify_crlbundle);
+s32 cert_get_signature(BbCertBase* certBase, void** out) { // TODO: rename cert_get_signature
+    switch (certBase->certType) {
+        case 0:
+            *out = &((BbEccCert*)certBase)->signature;
+            break;
+        case 1:
+            *out = &((BbRsaCert*)certBase)->signature;
+            break;
+        default:
+            return -1;
+    }
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", verify_all_crlbundles);
+    return 0;
```

Comment:
Could this return be duplicated into case 0 and 1? Would make it a bit more like the others so maybe preferable idk

Reply:
Unfortunately not. I also believe this is preferrable, as most functions tend to have the success path at the bottom of the file like this, while the failure paths are early returns. Yeah, the above functions are different in that regard, though I think they're a bit different in terms of "success" vs "failure"